Lib/fontTools/varLib/__init__.py

```diff
@@ -700,56 +676,141 @@ def load_designspace(designspace):
+			if axis.minimum != min(input_locations):
+				raise VarLibValidationError(f"Designspace {ds_path}: Axis '{axis.name}': the axis minimum value must be the lowest input value in the mapping.")
+			axis_min_output = axis.map_forward(axis.minimum)
+			if axis_min_output != min(output_locations):
+				raise VarLibValidationError(f"Designspace {ds_path}: Axis '{axis.name}': the axis minimum value must be the lowest output value in the mapping.")
+
+			if axis.default not in input_locations:
+				raise VarLibValidationError(f"Designspace {ds_path}: Axis '{axis.name}' has a mapping and must therefore include an entry for its own default value {axis.default} (<map input='{axis.default}' output='...'>).")
+
+			if axis.maximum not in input_locations:
+				raise VarLibValidationError(f"Designspace {ds_path}: Axis '{axis.name}' has a mapping and must therefore include an entry for its own maximum value {axis.maximum} (<map input='{axis.maximum}' output='...'>).")
+			if axis.maximum != max(input_locations):
+				raise VarLibValidationError(f"Designspace {ds_path}: Axis '{axis.name}': the axis maximum value must be the highest input value in the mapping.")
+			axis_max_output = axis.map_forward(axis.maximum)
+			if axis_max_output != max(output_locations):
+				raise VarLibValidationError(f"Designspace {ds_path}: Axis '{axis.name}': the axis maximum value must be the highest output value in the mapping.")
+
+			# No duplicate input or output values:
+			input_list = [input for input, _ in axis.map]
+			output_list = [output for _, output in axis.map]
+
+			if len(input_locations) != len(input_list):
+				raise VarLibValidationError(f"Designspace {ds_path}: Axis '{axis.name}': mapping input values must be unique but contain duplicates.")
+			if len(output_locations) != len(output_list):
+				raise VarLibValidationError(f"Designspace {ds_path}: Axis '{axis.name}': mapping output values must be unique but contain duplicates.")
+
+			# Input and output values are in ascending order:
+			if sorted(input_locations) != input_list:
+				raise VarLibValidationError(f"Designspace {ds_path}: Axis '{axis.name}': mapping input values must be in ascending order.")
+			if sorted(output_locations) != output_list:
+				raise VarLibValidationError(f"Designspace {ds_path}: Axis '{axis.name}': mapping output values must be in ascending order.")
+
+		axes[axis.name] = axis
 
-		axes[axis_name] = axis
 	log.info("Axes:\n%s", pformat([axis.asdict() for axis in axes.values()]))
 
 	# Check all master and instance locations are valid and fill in defaults
+	instances = ds.instances
 	for obj in masters+instances:
 		obj_name = obj.name or obj.styleName or ''
 		loc = obj.location
-		for axis_name in loc.keys():
-			assert axis_name in axes, "Location axis '%s' unknown for '%s'." % (axis_name, obj_name)
+		# NOTE: designspaceLib filters out axes in locations that are unknown, no need to check here.
```

Comment:
But is designspaceLib the only library that reads designspaces? I would say that the check can go away only if the designspace spec says that it's forbidden to mention locations along axes that are not defined previously?

Reply:
agree with jany, pls keep this assert as assert